### PR TITLE
Adds a simplified pretty printer for tree transformations

### PR DESCRIPTION
**Pull request by rbuckton**
Unfortunately, we cannot easily leverage our existing emitter to emit transformed nodes. The current emitter often assumes nodes are original source tree nodes, rather than synthesized nodes created during a transformation phase. Transformers are not required to set parent pointers on nodes, as they will often reuse existing source tree nodes and should not change the parent pointers on these nodes.

Instead we will add a new, simpler node printer whose responsibility is to emit the exact node tree provided.

The `printFiles` function will eventually supplant the existing `emitFiles` function. When called, this function will get a list of `Transformer` callbacks based on the provided compiler options, adding the node printer as the last transformer in the list.  This gives the node printer access to the `TransformationContext`, allowing it to process substitutions and track any hoisted declarations added during just-in-time substitution.

The node printer in `printFiles` is designed for performance, and minimizes the number of conditional branches within the emit for any node. The expectation is that all transformers have finished every non-substitution transformation before the printer is called, and therefore the node printer will print nodes verbatim.

In addition, similarly to how source maps were extracted from `emitter.ts` and into a `SourceMapWriter` to support the node printer, comment emit is now also extracted into a `CommentWriter`. 

**Related Pull Requests:**
* #6892 - Adds the visitor API foundation for tree transformations.
* #6983 - Adds the Transform Flags concept for tree transformations.
* #6986 - Adds the transformFiles API for tree transformations.
* #6987 - **Adds a simplified pretty printer for tree transformations.**
* #7191 - Adds the TypeScript transformer.
* #7192 - Adds the JSX transformer.
* #7193 - Adds the ES7 transformer.
* #7194 - Adds the ES6 transformer.
* #7206 - Adds the Module transformer.

[Review on Reviewable](https://reviewable.io/reviews/microsoft/typescript/6987)

### CODE REVIEW

**Comment by rbuckton** [comment]
Paging for review: @mhegazy, @ahejlsberg, @yuit, @DanielRosenwasser, @RyanCavanaugh, @vladima

**Comment by DanielRosenwasser** [comment]
image

image

**Comment by DanielRosenwasser** [comment]
You can get around this by using [Reviewable](https://reviewable.io/reviews/Microsoft/TypeScript/6987).

**Comment by DanielRosenwasser** [comment]
Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/compiler/comments.ts, line 143 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAClqdsq5r2BSWd-j38:-KAClqdsq5r2BSWd-j39:-108718151) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/comments.ts#L143)):*
emitted as

Put some sort of quote demarcation around the emit.

---

*[src/compiler/comments.ts, line 148 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KACmB6KotJ196jjohgm:-KACmB6KotJ196jjohgn:-2092293696) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/comments.ts#L148)):*
same here regarding demarcation and "as"

---

*[src/compiler/comments.ts, line 185 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KACmQ1GklzsAp9BGiay:-KACmQ1GklzsAp9BGiaz:-1980549181) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/comments.ts#L185)):*
Just use a numeric comparison

---

*[src/compiler/utilities.ts, line 115 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KACmqKCzcT_mPoow778:-KACmqKCzcT_mPoow779:82501046) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/utilities.ts#L115)):*
I thought it was CommonJS by default but this works too.

---

*[src/compiler/factory.ts, line 148 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KACnKr0zQ927qLebSyx:-KACnKr0zQ927qLebSyy:-1739514302) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/factory.ts#L148)):*
What's the point of this exactly? Why can't you say this just returns a `PrimaryExpression`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/typescript/6987)*

**Comment by rbuckton** [comment]
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/compiler/comments.ts, line 143 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAClqdsq5r2BSWd-j38:-KAfFmfonQ74olKLAHHO:1908144812) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/comments.ts#L143)):*
Copy/paste from emitter, but I'll take a look.

---

*[src/compiler/comments.ts, line 148 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KACmB6KotJ196jjohgm:-KAfFtnJFkqY_deYZHSd:1680066548) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/comments.ts#L148)):*
Copy/paste from emitter.

---

*[src/compiler/comments.ts, line 185 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KACmQ1GklzsAp9BGiay:-KANP7pbSJ15wrGbMnKm:542455468) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/comments.ts#L185)):*
Copy/paste from emitter. 

---

*[src/compiler/utilities.ts, line 115 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KACmqKCzcT_mPoow778:-KAgX7uwhDad4NMnW3pI:171363129) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/utilities.ts#L115)):*
Not when the target is `ES6`. It looks like someone else had the same idea though, so I will drop these in favor of `getEmitScriptTarget` and `getEmitModuleKind`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/typescript/6987)*

**Comment by DanielRosenwasser** [comment]
Review status: 0 of 7 files reviewed at latest revision, 13 unresolved discussions.

---

*[src/compiler/printer.ts, line 184 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAw8Ay49Lonvk_tAcd1:-KAw8Ay49Lonvk_tAcd2:1554496064) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L184)):*
Make it explicit with`TempFlags.Auto`

---

*[src/compiler/printer.ts, line 742 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAw8ObPuOqECFoWsxvY:-KAw8ObPuOqECFoWsxvZ:-1702949019) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L742)):*
Could we not emit a semicolon as part of emitting a signature list?

---

*[src/compiler/printer.ts, line 995 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAw9NaWtrq97y-h5zE9:-KAw9NaWtrq97y-h5zEA:-1029339402) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L995)):*
Does the pretty-printer make no assumptions about the structure? Is it the responsibility of the prior transformers to ensure that an object literal is not the immediate concise body?

---

*[src/compiler/printer.ts, line 1106 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAw9o2GQ8qfC8IyUZAK:-KAw9o2HVgquNKNuOtVb:-407219741) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L1106)):*
Why even do it with a prefix? Or if you do, why not just

```ts
emitExpressionWithPrefix(node.asteriskToken ? "yield* " : "yield ", node.expression);·
```

---

*[src/compiler/printer.ts, line 1788 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAwB8GOjRmE50GQOwaD:-KAwB8GOjRmE50GQOwaE:-518134039) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L1788)):*
Cache length

---

*[src/compiler/printer.ts, line 1798 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAwAoO1Rt6ZMEHhCp62:-KAwAoO1Rt6ZMEHhCp63:-1109790320) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L1798)):*
I don't like how the function doesn't document its return value. Can you make the function name more explicit and add a JSDoc comment?

---

*[src/compiler/printer.ts, line 1870 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAwB6Yjk33Bq4XhhSgE:-KAwB6Yjk33Bq4XhhSgF:-518134039) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L1870)):*
Cache length

---

*[src/compiler/printer.ts, line 1984 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAwB_Yykm5VWi-CNaqy:-KAwB_Yykm5VWi-CNaqz:1956537383) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L1984)):*
What is this condition for?

What about default initializers and optional parameters? Do we have tests for these cases?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/typescript/6987)*

**Comment by rbuckton** [comment]
Review status: 0 of 12 files reviewed at latest revision, 13 unresolved discussions.

---

*[src/compiler/printer.ts, line 742 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAw8ObPuOqECFoWsxvY:-KBATv_XhRek6LUe_6Li:-1813586253) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L742)):*
This code path won't be used until such a time as we migrate the declaration emitter to use the pretty printer, but in the declaration emitter we always emit a semicolon.

---

*[src/compiler/printer.ts, line 995 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAw9NaWtrq97y-h5zE9:-KBAUJnmff8_W1s8cPBm:-508369270) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L995)):*
That is correct, although I am adding some support to automatically parenthesize expressions to ensure the correct output. One reason for this is to make sure we have the right parentheses around nodes without having to know about the parent node during an individual transformation.

---

*[src/compiler/printer.ts, line 1106 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAw9o2GQ8qfC8IyUZAK:-KBAUdosJvrGp7nwcqy4:427815299) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L1106)):*
Because `yield;` is legal in JavaScript, and we don't want to emit `yield ;`.

---

*[src/compiler/printer.ts, line 1788 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAwB8GOjRmE50GQOwaD:-KBAUm6C1PzaBYPt_e37:-514344479) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L1788)):*
I've discussed this with @vladima in the past. V8 still does an array length lookup on every iteration to perform a bounds check, so caching the length doesn't necessarily have a real benefit.

---

*[src/compiler/printer.ts, line 1798 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAwAoO1Rt6ZMEHhCp62:-KBAVDUhKT55-vZqAyTC:-220024952) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L1798)):*
I'll fix this, though the equivalent function in `emitter.ts` is also undocumented.

---

*[src/compiler/printer.ts, line 1870 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAwB6Yjk33Bq4XhhSgE:-KBAVcUMMzN0e-YDZRtp:-1582418045) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L1870)):*
See above.

---

*[src/compiler/printer.ts, line 1984 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAwB_Yykm5VWi-CNaqy:-KBAVovbxc8eyWhFkwBF:-1346518307) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L1984)):*
These are effectively the same conditions as in `emitSignatureParametersForArrow` in `emitter.ts`, with some additional restrictions.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/typescript/6987)*

File: src/compiler/types.ts [hunk 1]
```diff
@@ -2793,14 +2796,16 @@ namespace ts {
...
+        EmitExportStar = 1 << 1,                 // The export * helper should be written to this node.
+        EmitSuperHelper = 1 << 2,                // Emit the basic _super helper for async methods.
+        EmitAdvancedSuperHelper = 1 << 3,        // Emit the advanced _super helper for async methods.
+        UMDDefine = 1 << 4,                      // This node should be replaced with the UMD define helper.
+        NoLexicalEnvironment = 1 << 5,           // A new LexicalEnvironment should *not* be introduced when emitting this node, this is primarily used when printing a SystemJS module.
+        SingleLine = 1 << 6,                     // The contents of this node should be emit on a single line.
+        AdviseOnEmitNode = 1 << 7,               // The node printer should invoke the onBeforeEmitNode and onAfterEmitNode callbacks when printing this node.
+        IsNotEmittedNode = 1 << 8,               // Is a node that is not emitted but whose comments should be preserved if possible.
+        EmitCommentsOfNotEmittedParent = 1 << 8, // Emits comments of missing parent nodes.
```
Comment (sandersn) [inline src/compiler/types.ts]: should be `1 << 9`

**Comment by sandersn** [comment]
Review status: 0 of 12 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[src/compiler/utilities.ts, line 2726 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBOe7Elmo2LFJE-avsg:-KBOe7Elmo2LFJE-avsh:1880926942) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/utilities.ts#L2726)):*
Shouldn't CollapseToStart set `end: range.pos` ?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/typescript/6987)*

**Comment by sandersn** [comment]
Reviewed 4 of 9 files at r1, 1 of 9 files at r7, 3 of 5 files at r12, 1 of 1 files at r13.
Review status: 6 of 12 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[src/compiler/emitter.ts, line 2241 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBOjaYrPNm5WclL79EQ:-KBOjaYsH78WO-YkLttz:1151967650) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/emitter.ts#L2241)):*
After the split into `skipAssertions`, `skipParentheses` still skips type assertions and as expressions. Looks like it should probably just skip parenthesized expressions.

---

*[src/compiler/emitter.ts, line 4337 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBOkfHMWGUpFoJsXLf7:-KBOkfHN1h3wgZm6xiFT:-641526981) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/emitter.ts#L4337)):*
`transformCompatibleEmit` is temporary for backward compatibility with the old emitter right? I find it a lot easier to read this condition as `restIndex === 0 && compilerOptions.transformCompatibleEmit` -- I spent too long trying wondering why the new version went out of its way to avoid emitting a harmless `- 0`.

---

*[src/compiler/utilities.ts, line 2300 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBOfvsaq50oHGcWe0OK:-KBOfvsbZrgkQTXt4bVu:-106015148) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/utilities.ts#L2300)):*
typo:Separator

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/typescript/6987)*

**Comment by sandersn** [comment]
Reviewed 2 of 9 files at r1, 1 of 3 files at r5, 1 of 1 files at r9.
Review status: 7 of 12 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[src/compiler/printer.ts, line 1984 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAwB_Yykm5VWi-CNaqy:-KBPk_o2HVk7whuybUOz:-1562018295) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L1984)):*
is `parameters[0].type === undefined` needed, though? I don't think there's a way to get `parameters[0].pos === parentNode.pos` if there's a type annotation.

---

*[src/compiler/printer.ts, line 1568 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBPVnvSzEu44JRO25Cn:-KBPVnvSzEu44JRO25Co:-1816183512) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/printer.ts#L1568)):*
Why is this special-cased?

---

*[src/compiler/printer.ts, line 1723 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBPavTiwyCseOyNO9sA:-KBPavTiwyCseOyNO9sB:-1558694874) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/printer.ts#L1723)):*
debugger

---

*[src/compiler/printer.ts, line 1773 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBPbWftcxl_Bkt5MNQR:-KBPbWftcxl_Bkt5MNQS:1152690036) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/printer.ts#L1773)):*
is this comment current? What about the commented-out line?

---

*[src/compiler/printer.ts, line 1958 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBPiy8A0QoQHh086Zv3:-KBPiy8A0QoQHh086Zv4:-1036476716) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/printer.ts#L1958)):*
why not just check that `modifiers && modifiers.length` ?

---

*[src/compiler/printer.ts, line 2349 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBPohPDvGW7G5eu1k__:-KBPohPDvGW7G5eu1k_a:-539164917) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/printer.ts#L2349)):*
This check seems very specific to const enums. Maybe it would be worthwhile to split into two functions or just inline this check into `tryEmitConstantValue`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/typescript/6987)*

**Comment by sandersn** [comment]
Besides @DanielRosenwasser and my questions and the small suggestion on line 2349, :+1:

**Comment by rbuckton** [comment]
Review status: 7 of 12 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[src/compiler/comments.ts, line 143 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAClqdsq5r2BSWd-j38:-KBigUd6aP3ausGdDkk-:-499146009) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/comments.ts#L143)):*
Done.

---

*[src/compiler/comments.ts, line 148 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KACmB6KotJ196jjohgm:-KBigV5KH6HyBfxc4fD7:-499146009) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/comments.ts#L148)):*
Done.

---

*[src/compiler/comments.ts, line 185 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KACmQ1GklzsAp9BGiay:-KBigW6A-bJtLMWunVXS:-499146009) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/comments.ts#L185)):*
Done.

---

*[src/compiler/emitter.ts, line 2241 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBOjaYrPNm5WclL79EQ:-KBigX88ZtqDW7p8STIo:1478449924) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/emitter.ts#L2241)):*
Where `skipParentheses` is called it expects to skip past all parentheses *and* assertions, so I'll rename the method.

---

*[src/compiler/emitter.ts, line 4337 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBOkfHMWGUpFoJsXLf7:-KBigh6RXPMAmD5lV3XY:1061752228) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/emitter.ts#L4337)):*
I am going to remove `transformCompatibleEmit` and make it the default, it will help with testing.

---

*[src/compiler/printer.ts, line 184 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAw8Ay49Lonvk_tAcd1:-KBigsfPqesqF5p5Y7Qo:-499146009) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L184)):*
Done.

---

*[src/compiler/printer.ts, line 1984 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KAwB_Yykm5VWi-CNaqy:-KBij8hFuMcJc9SssXm4:751740795) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/printer.ts#L1984)):*
The exhaustive check is due to the fact you could, in theory, create a synthesized arrow function with a parameter that has a type annotation.

---

*[src/compiler/printer.ts, line 1568 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBPVnvSzEu44JRO25Cn:-KBih2lTGOQzcQB3yLDU:1686152322) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/printer.ts#L1568)):*
We need to do different things when we emit an `Identifier` that is not an expressions. In `emitter.ts` we traverse parent pointers to discover if an `Identifier` is an expression, but we cannot use parent pointers in tree transforms. Instead, when we print something in an expression position we call `emitExpression` rather than `emit`. This allows us to accomplish the same thing without the parent traversal.

---

*[src/compiler/printer.ts, line 1723 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBPavTiwyCseOyNO9sA:-KBihj3kW4xuLR9FpMHJ:-499146009) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/printer.ts#L1723)):*
Done.

---

*[src/compiler/printer.ts, line 1773 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBPbWftcxl_Bkt5MNQR:-KBiiQWbea8QjWfhPNQm:-505686368) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/printer.ts#L1773)):*
This was not completed work. I'll take a look.

---

*[src/compiler/printer.ts, line 1958 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBPiy8A0QoQHh086Zv3:-KBij4GK_xRBIHtel83w:-499146009) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/printer.ts#L1958)):*
Done.

---

*[src/compiler/printer.ts, line 2349 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBPohPDvGW7G5eu1k__:-KBijdowYSeXoBjq4Nh5:2073966726) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/printer.ts#L2349)):*
No, this is not specific to const enums. `getTextOfNode` comes from `emitter.ts` and is used in the same fashion, however I needed to add this test to catch synthesized nodes that do not correspond to a source location.

---

*[src/compiler/types.ts, line 2808 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KACkn0DOapLzRDc7mEt-r14-2808:-KBilQq1mz9CfwPqeIEi:-499146009) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/types.ts#L2808)):*
Done.

---

*[src/compiler/utilities.ts, line 2300 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBOfvsaq50oHGcWe0OK:-KBilRvLgnRuGsp9LWD4:-499146009) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/utilities.ts#L2300)):*
Done.

---

*[src/compiler/utilities.ts, line 2726 \[r14\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KBOe7Elmo2LFJE-avsg:-KBilS_tno4tT-q2lXiz:-499146009) ([raw file](https://github.com/microsoft/typescript/blob/75b2181e17c88d62b9a17f1d25afb05e2a3e42f2/src/compiler/utilities.ts#L2726)):*
Done.

---

*[src/compiler/factory.ts, line 148 \[r1\]](https://reviewable.io:443/reviews/microsoft/typescript/6987#-KACnKr0zQ927qLebSyx:-KBilTbrOuFhS4FfJv83:-499146009) ([raw file](https://github.com/microsoft/typescript/blob/6b381ecdbd81e47ec122a493d27529dd36db20b2/src/compiler/factory.ts#L148)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/typescript/6987)*

**Pull request merged**